### PR TITLE
Add mapToResult operator, tests

### PR DESCRIPTION
**Pull request by kaphacius**
This PR adds a new operator `catchToResult` which allows transforming a `Publisher` that has a specific `Failure` type (not `Never`) into a publisher that outputs `Output` and `Failure` wrapped in a `Result`. 
A common use-case for this is when an interface should produce publishers of type `AnyPublisher<Result<T, Error>, Never>`, but the underlying code works with publishers like `AnyPublisher<T, Error>` (`URLSession.DataTaskPublisher`).

The opposite situation, going from `AnyPublisher<Result<T, Error>, Never>` to `AnyPublisher<T, Error>` also happens. If there is an interest in this PR I can open another one with an operator that allows doing that.

### CODE REVIEW

File: Sources/Operators/CatchToResult.swift [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+//
+//  Created by Yurii Zadoianchuk on 05/03/2021.
+//  Copyright © 2021 Combine Community. All rights reserved.
+//
+
+#if canImport(Combine)
+import Combine
+
+@available(OSX 10.15, iOS 13.0, tvOS 13.0, watchOS 6.0, *)
+public extension Publisher where Failure: Error {
```
Comment (manmal) [inline Sources/Operators/CatchToResult.swift:13]: I think `Failure` always needs to conform to `Error`, so `where Failure: Error` could probably be removed?

File: Sources/Operators/CatchToResult.swift [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+    /// Transform a publisher with concrete Output and Failure types
+    /// to a new publisher that wraps Output and Failure in Result,
+    /// and has Never for Failure type
+    /// - Returns: A type-erased publiser of type <Result<Output, Failure>, Never>
+    func catchToResult() -> AnyPublisher<Result<Output, Failure>, Never> {
+        self.map({ Result.success($0) })
+            .catch({ Just(.failure($0)) })
+            .eraseToAnyPublisher()
```
Comment (manmal) [inline Sources/Operators/CatchToResult.swift:21]: Nitpick: `self` isn't necessary here, and braces can be simplified:
```swift
map(Result.success)
    .catch { Just(.failure($0)) }
    .eraseToAnyPublisher()
```

**Review by manmal — COMMENTED** [review]
I think this looks great, thanks for taking up the idea!

**Comment by jasdev** [comment]
Would love to see this land, thanks @kaphacius! I use TCA’s [`catchToEffect` analog](https://github.com/pointfreeco/swift-composable-architecture/blob/e93527097282d39588d1d4c897b9323301785974/Sources/ComposableArchitecture/Effect.swift#L319-L335) often. ✅

> The opposite situation, going from `AnyPublisher<Result<Output, Failure>, Never>` to AnyPublisher<Output, Failure> also happens. If there is an interest in this PR I can open another one with an operator that allows doing that.

+1.

(and as a heads up, we’ll probs need some quick docs. in the README.md, for when you get a chance. 🙏🏽)

File: Tests/CatchToResultTests.swift [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+
+    func testCatchCustomError() {
+        let subject = PassthroughSubject<Int, Error>()
+        var completed = false
+        var gotFailure = false
+        var gotSuccess = false
+        var result: Result<Int, Error>? = nil
+
+        subscription = subject
+            .tryMap({ _ -> Int in throw(CatchToResultError.someError) })
```
Comment (freak4pc) [inline Tests/CatchToResultTests.swift:53]:
```suggestion
            .tryMap { _ -> Int in throw CatchToResultError.someError }
```

please drop parentheses in trailing-closures everywhere

Also, throw isn't a function, it's a specialty word.

File: Tests/CatchToResultTests.swift [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+
+@available(OSX 10.15, iOS 13.0, tvOS 13.0, watchOS 6.0, *)
+final class CatchToResultTests: XCTestCase {
+    private var subscription: AnyCancellable!
+
+    enum CatchToResultError: Error {
+        case someError
+    }
+
+    func testCatchToResult() {
```
Comment (freak4pc) [inline Tests/CatchToResultTests.swift:24]:
```suggestion
    func testCatchResultNoError() {
```

**Review by freak4pc — COMMENTED** [review]
Thanks! Tests look good overall, but there are some formatting issues, for example usage of unneeded parentheses, using keywords like `catch` and `throw` as functions (e.g. `catch (let e)`, `throw(MyError.error)`. 

If you'll clean up these things I'll be happy to merge.

Thank you!

**Comment by kaphacius** [comment]
Added info to README.md and applied all of the suggested comments 👍

**Comment by jshier** [comment]
Is there any particular reason not to make the name of this something more general, like `toResult`? It's not strictly a catching operator, as it transforms the success case as well. 

(I have this operator in my code base as `asResult()`, though I'm not particular on `as` vs. `to`.)

**Comment by freak4pc** [comment]
I agree about @jshier's remarks and this is the biggest gripe I have here. Something like maoToResult or something else makes more sense, but there's "prior art" with TCA's catchToEffect. Not sure that's really a strong enough reason to not give it our own naming.

**Comment by kaphacius** [comment]
I don't have any attachment to the name to be honest, I think `mapToResult` sounds equally good. I'll make the changes and update the PR

**Comment by kaphacius** [comment]
Renamed the operator, tests and the PR name :)

**Review by freak4pc — APPROVED** [review]
Looks great! Thank you :)

**Pull request merged**